In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv')
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [3]:
df.shape

(7043, 21)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [5]:
df.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

In [6]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [7]:
df.duplicated().sum()

np.int64(0)

First, I'll convert the 'TotalCharges' column to a numeric type

In [8]:
# Some values in 'TotalCharges' are not numeric, so I'll coerce errors to NaN
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

# Then, I'll fill any missing values in 'TotalCharges' with 0
df['TotalCharges'] = df['TotalCharges'].fillna(0)

# Verify the conversion and filling
print(df['TotalCharges'].info())
print(df['TotalCharges'].isnull().sum())

<class 'pandas.core.series.Series'>
RangeIndex: 7043 entries, 0 to 7042
Series name: TotalCharges
Non-Null Count  Dtype  
--------------  -----  
7043 non-null   float64
dtypes: float64(1)
memory usage: 55.2 KB
None
0


then i will Convert the 'Churn' column to binary values, where 'No' is mapped to 0 and 'Yes' is mapped to 1.


In [9]:
#then i will Convert the 'Churn' column to binary values, where 'No' is mapped to 0 and 'Yes' is mapped to 1.

df['Churn'] = df['Churn'].map({'No': 0, 'Yes': 1})

In [10]:
df['Churn']

0       0
1       0
2       1
3       0
4       1
       ..
7038    0
7039    0
7040    0
7041    1
7042    0
Name: Churn, Length: 7043, dtype: int64

In [12]:
df.head(

)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges  Churn  
0           Electronic check          29.85         29.85      0  
1               Mailed check          56.95       1889.50      0  
2               Mailed check          53.85        108.15      1  
3  Bank transfer (automatic)          42.30       1840.75      0  
4           Electronic check          70.70        151.65      1  

[5 rows x 21 columns]

Split the data into an 80-20 train-test split with a random state of “1”.
Using this features: categorical = ['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService',
                'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
                'Contract', 'PaperlessBilling', 'PaymentMethod']
 numerical = ['tenure', 'MonthlyCharges', 'TotalCharges']

In [14]:
from sklearn.model_selection import train_test_split

categorical = ['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService',
               'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
               'Contract', 'PaperlessBilling', 'PaymentMethod']
numerical = ['tenure', 'MonthlyCharges', 'TotalCharges']
features = categorical + numerical
target = 'Churn'

X = df[features]
y = df[target]
#generated from google colab

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1, stratify=y)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (5634, 19)
Shape of X_test: (1409, 19)
Shape of y_train: (5634,)
Shape of y_test: (1409,)


The numerical features is scaled using StandardScaler, converted the output back to a dataframe and put back the column names.


In [21]:
#generated from google colab
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_numerical_scaled = scaler.fit_transform(X_train[numerical])
X_test_numerical_scaled = scaler.transform(X_test[numerical])

X_train[numerical] = pd.DataFrame(X_train_numerical_scaled, columns=numerical, index=X_train.index)
X_test[numerical] = pd.DataFrame(X_test_numerical_scaled, columns=numerical, index=X_test.index)

X_train.head()

gender  SeniorCitizen Partner Dependents PhoneService MultipleLines  \
2246  Female              0     Yes        Yes          Yes           Yes   
6843  Female              0     Yes         No          Yes           Yes   
159     Male              0      No         No          Yes           Yes   
6259    Male              0     Yes        Yes          Yes            No   
4291  Female              0     Yes        Yes          Yes           Yes   

     InternetService       OnlineSecurity         OnlineBackup  \
2246     Fiber optic                  Yes                   No   
6843     Fiber optic                  Yes                  Yes   
159               No  No internet service  No internet service   
6259             DSL                   No                  Yes   
4291     Fiber optic                  Yes                  Yes   

         DeviceProtection          TechSupport          StreamingTV  \
2246                  Yes                   No                  Yes   
6843                  Yes                   No                   No   
159   No internet service  No internet service  No internet service   
6259                   No                  Yes                  Yes   
4291                   No                  Yes                  Yes   

          StreamingMovies        Contract PaperlessBilling  \
2246                  Yes  Month-to-month              Yes   
6843                   No  Month-to-month               No   
159   No internet service        Two year               No   
6259                  Yes        Two year               No   
4291                  Yes        Two year               No   

                  PaymentMethod    tenure  MonthlyCharges  TotalCharges  
2246           Electronic check -1.274670        1.252068     -0.959674  
6843  Bank transfer (automatic) -0.949139        0.895785     -0.622602  
159   Bank transfer (automatic) -0.908448       -1.303514     -0.893814  
6259  Bank transfer (automatic)  1.126118        0.402982      1.001558  
4291  Bank transfer (automatic)  1.288883        1.513454      2.084884

In [23]:
# prompted and generated from google colab : The categorical features are one-hot encoded using OneHotEncoder(set sparse_output to false), convert the output back to a dataframe and put back the column names.

from sklearn.preprocessing import OneHotEncoder

ohe = OneHotEncoder(sparse_output=False)

X_train_categorical_ohe = ohe.fit_transform(X_train[categorical])
X_test_categorical_ohe = ohe.transform(X_test[categorical])

# Create DataFrames for the one-hot encoded categorical features
X_train_categorical_ohe_df = pd.DataFrame(X_train_categorical_ohe, columns=ohe.get_feature_names_out(categorical), index=X_train.index)
X_test_categorical_ohe_df = pd.DataFrame(X_test_categorical_ohe, columns=ohe.get_feature_names_out(categorical), index=X_test.index)

# Concatenate the one-hot encoded categorical features with the scaled numerical features
X_train = pd.concat([X_train_categorical_ohe_df, X_train[numerical]], axis=1)
X_test = pd.concat([X_test_categorical_ohe_df, X_test[numerical]], axis=1)

X_train.head()

gender_Female  gender_Male  SeniorCitizen_0  SeniorCitizen_1  \
2246            1.0          0.0              1.0              0.0   
6843            1.0          0.0              1.0              0.0   
159             0.0          1.0              1.0              0.0   
6259            0.0          1.0              1.0              0.0   
4291            1.0          0.0              1.0              0.0   

      Partner_No  Partner_Yes  Dependents_No  Dependents_Yes  PhoneService_No  \
2246         0.0          1.0            0.0             1.0              0.0   
6843         0.0          1.0            1.0             0.0              0.0   
159          1.0          0.0            1.0             0.0              0.0   
6259         0.0          1.0            0.0             1.0              0.0   
4291         0.0          1.0            0.0             1.0              0.0   

      PhoneService_Yes  ...  Contract_Two year  PaperlessBilling_No  \
2246               1.0  ...                0.0                  0.0   
6843               1.0  ...                0.0                  1.0   
159                1.0  ...                1.0                  1.0   
6259               1.0  ...                1.0                  1.0   
4291               1.0  ...                1.0                  1.0   

      PaperlessBilling_Yes  PaymentMethod_Bank transfer (automatic)  \
2246                   1.0                                      0.0   
6843                   0.0                                      1.0   
159                    0.0                                      1.0   
6259                   0.0                                      1.0   
4291                   0.0                                      1.0   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
2246                                    0.0                             1.0   
6843                                    0.0                             0.0   
159                                     0.0                             0.0   
6259                                    0.0                             0.0   
4291                                    0.0                             0.0   

      PaymentMethod_Mailed check    tenure  MonthlyCharges  TotalCharges  
2246                         0.0 -1.274670        1.252068     -0.959674  
6843                         0.0 -0.949139        0.895785     -0.622602  
159                          0.0 -0.908448       -1.303514     -0.893814  
6259                         0.0  1.126118        0.402982      1.001558  
4291                         0.0  1.288883        1.513454      2.084884  

[5 rows x 46 columns]

In [24]:
# prompted on google colab: Combine scaled numerical and one-hot encoded categorical features into train and test set dataframes (use pd.concat)

# Now, concatenate the scaled numerical features and the one-hot encoded categorical features
X_train_processed = pd.concat([X_train[numerical], X_train_categorical_ohe_df], axis=1)
X_test_processed = pd.concat([X_test[numerical], X_test_categorical_ohe_df], axis=1)

print("Shape of processed X_train:", X_train_processed.shape)
print("Shape of processed X_test:", X_test_processed.shape)

X_train_processed.head()
X_test_processed.head()


Shape of processed X_train: (5634, 46)
Shape of processed X_test: (1409, 46)


tenure  MonthlyCharges  TotalCharges  gender_Female  gender_Male  \
6826 -1.152596       -1.470001     -0.967932            0.0          1.0   
687  -1.233978       -0.519358     -0.964686            0.0          1.0   
3787  1.044736        1.013991      1.564008            1.0          0.0   
4130 -0.908448       -0.952225     -0.839766            1.0          0.0   
640   1.451649       -1.313503     -0.289482            0.0          1.0   

      SeniorCitizen_0  SeniorCitizen_1  Partner_No  Partner_Yes  \
6826              1.0              0.0         1.0          0.0   
687               0.0              1.0         1.0          0.0   
3787              1.0              0.0         1.0          0.0   
4130              1.0              0.0         1.0          0.0   
640               1.0              0.0         0.0          1.0   

      Dependents_No  ...  StreamingMovies_Yes  Contract_Month-to-month  \
6826            1.0  ...                  0.0                      1.0   
687             1.0  ...                  0.0                      1.0   
3787            1.0  ...                  1.0                      1.0   
4130            1.0  ...                  0.0                      1.0   
640             0.0  ...                  0.0                      0.0   

      Contract_One year  Contract_Two year  PaperlessBilling_No  \
6826                0.0                0.0                  0.0   
687                 0.0                0.0                  0.0   
3787                0.0                0.0                  0.0   
4130                0.0                0.0                  1.0   
640                 0.0                1.0                  1.0   

      PaperlessBilling_Yes  PaymentMethod_Bank transfer (automatic)  \
6826                   1.0                                      0.0   
687                    1.0                                      0.0   
3787                   1.0                                      0.0   
4130                   0.0                                      0.0   
640                    0.0                                      1.0   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
6826                                    1.0                             0.0   
687                                     0.0                             1.0   
3787                                    0.0                             1.0   
4130                                    0.0                             1.0   
640                                     0.0                             0.0   

      PaymentMethod_Mailed check  
6826                         0.0  
687                          0.0  
3787                         0.0  
4130                         0.0  
640                          0.0  

[5 rows x 46 columns]

prompted from google colab

In [25]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Initialize and train the Random Forest Classifier
rf_model = RandomForestClassifier(random_state=1)
rf_model.fit(X_train_processed, y_train)

print("Random Forest Classifier trained successfully!")

Random Forest Classifier trained successfully!


In [26]:
# Predict on the test set
y_pred_rf = rf_model.predict(X_test_processed)

In [28]:
# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred_rf))

Accuracy: 0.7863733144073811


In [29]:
print("Classification Report:", classification_report(y_test, y_pred_rf))

Classification Report:               precision    recall  f1-score   support

           0       0.83      0.89      0.86      1035
           1       0.62      0.49      0.55       374

    accuracy                           0.79      1409
   macro avg       0.73      0.69      0.70      1409
weighted avg       0.77      0.79      0.78      1409



In [30]:
print("Confusion Matrix:", confusion_matrix(y_test, y_pred_rf))

Confusion Matrix: [[925 110]
 [191 183]]


In [31]:
from sklearn.ensemble import ExtraTreesClassifier

# Initialize and train the Extra Trees Classifier
et_model = ExtraTreesClassifier(random_state=1)
et_model.fit(X_train_processed, y_train)

ExtraTreesClassifier(random_state=1)

In [32]:
# Predict on the test set
y_pred_et = et_model.predict(X_test_processed)

# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred_et))

Accuracy: 0.7629524485450674


In [33]:
print("\nClassification Report:\n", classification_report(y_test, y_pred_et))


Classification Report:
               precision    recall  f1-score   support

           0       0.82      0.87      0.84      1035
           1       0.56      0.47      0.51       374

    accuracy                           0.76      1409
   macro avg       0.69      0.67      0.68      1409
weighted avg       0.75      0.76      0.76      1409



In [34]:
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_et))


Confusion Matrix:
 [[899 136]
 [198 176]]


In [35]:
import xgboost as xgb

# Initialize and train the XGBoost Classifier
xgb_model = xgb.XGBClassifier(random_state=1)
xgb_model.fit(X_train_processed, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, random_state=1, ...)

In [36]:
# Predict on the test set
y_pred_xgb = xgb_model.predict(X_test_processed)

# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred_xgb))

Accuracy: 0.78708303761533


In [37]:
print("\nClassification Report:\n", classification_report(y_test, y_pred_xgb))


Classification Report:
               precision    recall  f1-score   support

           0       0.83      0.89      0.86      1035
           1       0.62      0.51      0.56       374

    accuracy                           0.79      1409
   macro avg       0.73      0.70      0.71      1409
weighted avg       0.78      0.79      0.78      1409



In [38]:
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_xgb))


Confusion Matrix:
 [[918 117]
 [183 191]]


In [39]:
import lightgbm as lgb

# Initialize and train the LightGBM Classifier
lgb_model = lgb.LGBMClassifier(random_state=1)
lgb_model.fit(X_train_processed, y_train)

[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Number of positive: 1495, number of negative: 4139
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001656 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 670
[LightGBM] [Info] Number of data points in the train set: 5634, number of used features: 46
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.265353 -> initscore=-1.018328
[LightGBM] [Info] Start training from score -1.018328


LGBMClassifier(random_state=1)

In [40]:
# Predict on the test set
y_pred_lgb = lgb_model.predict(X_test_processed)

# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred_lgb))

Accuracy: 0.8005677785663591


In [41]:
print("\nClassification Report:\n", classification_report(y_test, y_pred_lgb))


Classification Report:
               precision    recall  f1-score   support

           0       0.84      0.90      0.87      1035
           1       0.65      0.53      0.59       374

    accuracy                           0.80      1409
   macro avg       0.75      0.71      0.73      1409
weighted avg       0.79      0.80      0.79      1409



In [42]:
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_lgb))


Confusion Matrix:
 [[929 106]
 [175 199]]


Here is a summary table of the performance metrics for each model:

| Model                 | Accuracy | Precision (Churn=1) | Recall (Churn=1) | F1-score (Churn=1) | Confusion Matrix (True Neg, False Pos, False Neg, True Pos) |
|-----------------------|----------|---------------------|------------------|--------------------|-------------------------------------------------------------|
| Random Forest         | 0.786    | 0.62                | 0.49             | 0.55               | [[925, 110], [191, 183]]                                    |
| Extra Trees           | 0.763    | 0.56                | 0.47             | 0.51               | [[899, 136], [198, 176]]                                    |
| XGBoost               | 0.787    | 0.62                | 0.51             | 0.56               | [[918, 117], [183, 191]]                                    |
| LightGBM              | 0.801    | 0.65                | 0.53             | 0.59               | [[929, 106], [175, 199]]                                    |



question 3

𝐹
1
=
2
⋅
Precision
⋅
Recall
Precision
+
Recall
Where:

Precision
=
True Positives
True Positives
+
False Positives
=
1380
1380
+
255
Recall
=
True Positives
True Positives
+
False Negatives
=
1380
1380
+
20
Plugging in the values:

Precision
≈
0.844
Recall
≈
0.9857
Now, calculating the F1 score:

𝐹
1
≈
2
⋅
0.844
⋅
0.9857
0.844
+
0.9857
≈
0.9093

In [43]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import ExtraTreesClassifier

# Define the hyperparameter grid
n_estimators = [50, 100, 300, 500, 1000]
min_samples_split = [2, 3, 5, 7, 9]
min_samples_leaf = [1, 2, 4, 6, 8]
max_features = ['sqrt', 'log2', None] # 'auto' is deprecated, use 'sqrt' or 'log2'

hyperparameter_grid = {'n_estimators': n_estimators,
                       'min_samples_leaf': min_samples_leaf,
                       'min_samples_split': min_samples_split,
                       'max_features': max_features}

# Initialize the Extra Trees Classifier
et_model = ExtraTreesClassifier(random_state=1)

# Initialize RandomizedSearchCV
random_search = RandomizedSearchCV(estimator=et_model,
                                   param_distributions=hyperparameter_grid,
                                   cv=5,
                                   n_iter=10,
                                   scoring='accuracy',
                                   n_jobs=-1,
                                   verbose=1,
                                   random_state=1)

# Perform the randomized search
random_search.fit(X_train_processed, y_train)

# Print the best hyperparameters
print("Best Hyperparameters:", random_search.best_params_)

Fitting 5 folds for each of 10 candidates, totalling 50 fits
Best Hyperparameters: {'n_estimators': 50, 'min_samples_split': 7, 'min_samples_leaf': 6, 'max_features': 'sqrt'}


In [44]:
# Train a new Extra Trees Classifier with the best hyperparameters
best_et_model = ExtraTreesClassifier(n_estimators=random_search.best_params_['n_estimators'],
                                      min_samples_leaf=random_search.best_params_['min_samples_leaf'],
                                      min_samples_split=random_search.best_params_['min_samples_split'],
                                      max_features=random_search.best_params_['max_features'],
                                      random_state=1)

best_et_model.fit(X_train_processed, y_train)

# Predict on the test set with the new model
y_pred_best_et = best_et_model.predict(X_test_processed)

# Evaluate the new model
accuracy_best_et = accuracy_score(y_test, y_pred_best_et)
print("Accuracy of the Extra Trees Classifier with best hyperparameters:", accuracy_best_et)

# Compare with the initial Extra Trees Classifier accuracy (from the previous output)
# Initial Extra Trees Accuracy: 0.7629524485450674

initial_et_accuracy = 0.7629524485450674  # Replace with the actual initial accuracy if needed

if accuracy_best_et > initial_et_accuracy:
    print("\nThe accuracy of the new optimal model is higher than the initial ExtraTreesClassifier model.")
elif accuracy_best_et < initial_et_accuracy:
    print("\nThe accuracy of the new optimal model is lower than the initial ExtraTreesClassifier model.")
else:
    print("\nThe accuracy of the new optimal model is the same as the initial ExtraTreesClassifier model.")

Accuracy of the Extra Trees Classifier with best hyperparameters: 0.794180269694819

The accuracy of the new optimal model is higher than the initial ExtraTreesClassifier model.
